etc/redux-toolkit.api.md

```diff
@@ -18,32 +18,61 @@ import { Store } from 'redux';
+}
+
 // Warning: (ae-missing-release-tag) "ActionCreatorWithOptionalPayload" is exported by the package, but it is missing a release tag (@alpha, @beta, @public, or @internal)
 //
 // @public (undocumented)
-export type ActionCreatorWithOptionalPayload<P, T extends string = string> = WithTypePropertyAndMatch<{
+export interface ActionCreatorWithOptionalPayload<P, T extends string = string> extends BaseActionCreator<P, T> {
+    // (undocumented)
     (payload?: undefined): PayloadAction<undefined, T>;
+    // Warning: (ae-forgotten-export) The symbol "Diff" needs to be exported by the entry point index.d.ts
+    //
+    // (undocumented)
     <PT extends Diff<P, undefined>>(payload?: PT): PayloadAction<PT, T>;
-}, T, P | undefined>;
+}
 
 // Warning: (ae-missing-release-tag) "ActionCreatorWithoutPayload" is exported by the package, but it is missing a release tag (@alpha, @beta, @public, or @internal)
 //
 // @public (undocumented)
-export type ActionCreatorWithoutPayload<T extends string = string> = WithTypePropertyAndMatch<() => PayloadAction<undefined, T>, T, undefined>;
+export interface ActionCreatorWithoutPayload<T extends string = string> extends BaseActionCreator<undefined, T> {
+    // (undocumented)
+    (): PayloadAction<undefined, T>;
+}
 
-// Warning: (ae-forgotten-export) The symbol "IsUnknownOrNonInferrable" needs to be exported by the entry point index.d.ts
 // Warning: (ae-missing-release-tag) "ActionCreatorWithPayload" is exported by the package, but it is missing a release tag (@alpha, @beta, @public, or @internal)
 //
 // @public (undocumented)
-export type ActionCreatorWithPayload<P, T extends string = string> = WithTypePropertyAndMatch<IsUnknownOrNonInferrable<P, <PT extends unknown>(payload: PT) => PayloadAction<PT, T>, <PT extends P>(payload: PT) => PayloadAction<PT, T>>, T, P>;
+export interface ActionCreatorWithPayload<P, T extends string = string> extends BaseActionCreator<P, T> {
+    // (undocumented)
+    <PT extends P>(payload: PT): PayloadAction<PT, T>;
+    // (undocumented)
+    (payload: P): PayloadAction<P, T>;
+}
 
-// Warning: (ae-forgotten-export) The symbol "MetaOrNever" needs to be exported by the entry point index.d.ts
-// Warning: (ae-forgotten-export) The symbol "ErrorOrNever" needs to be exported by the entry point index.d.ts
 // Warning: (ae-missing-release-tag) "ActionCreatorWithPreparedPayload" is exported by the package, but it is missing a release tag (@alpha, @beta, @public, or @internal)
 //
 // @public (undocumented)
-export type ActionCreatorWithPreparedPayload<PA extends PrepareAction<any> | void, T extends string = string> = PA extends PrepareAction<infer P> ? WithTypePropertyAndMatch<(...args: Parameters<PA>) => PayloadAction<P, T, MetaOrNever<PA>, ErrorOrNever<PA>>, T, P, MetaOrNever<PA>, ErrorOrNever<PA>> : void;
+export interface ActionCreatorWithPreparedPayload<Args extends unknown[], P, T extends string = string, E = never, M = never> extends BaseActionCreator<P, T, M, E> {

```

Comment:
Renamed type `ActionCreatorWithPreparedPayload` to `_ActionCreatorWithPreparedPayload`. This type had a signature that made it almost unreadable for external use, but very handy for internal use.

Created new interface with the old name `ActionCreatorWithPreparedPayload`, now with a more readable signature. The original renamed type essentially just redirects to this new type.

Reply:
@markerikson I've thought about this one, and given it's the only breaking typechange, I'd rather like to avoid that.
An alternative would be to keep the type `ActionCreatorWithPreparedPayload` with the existing signature instead of renaming it and introducing a new interface to be exposed to the user.
But I'd need a new name for that and I'm not 100% sure what would be fitting. Something along the lines of `PreparedActionCreator` or `EnhancedActionCreator`. What do you think?

Reply:
So this is where we get into the question of "what really is a breaking change, anyway?".

Is it a literal change to an API definition that is different than how the old one behaved?  Is it something that causes runtime errors?  Is it something that causes a compilation failure?

One of the downsides I've observed with TS is that almost _any_ change _can_ result in a compile error, even for stuff that's only intended to fix some bugs.  Do those count as "breaking changes"?

Given that, I am inclined to treat stuff like this as a patch or a minor release, rather than a semver major.  We've identified an issue with the types that doesn't actually affect the rest of the functionality, and we're trying to resolve that issue.  

If you're worried about folks relying on that particular type, a quick search on Github turns up exactly 3 hits: 2 in RTK, and 1 where someone copy-pasted some RTK code.  So I don't expect this will suddenly break people.

Reply:
Yup, that stuff gets pretty philosophical pretty fast.

In this case, I guess it's okay since the original type was *very* unlikely to be used by anyone (it was very unwieldy and I don't see a reason why someone would want to manually specify it anyways), as your github search confirms.

In general, I guess we'll do good to keep those type APIs as stable as possible, but that's what we've got these reports for, going forward from now. I guess people are somehow used to types breaking, but we'd probably do good to handle type-breaking changes as at least minor releases - I guess it should be possible to apply patch-level upgrades without second thought or consequence.

Reply:
Yeah, I'm on board with that.  So, this would end up as 1.2.0.

I certainly don't want to go randomly breaking types, but I do feel that types changes should be treated with a bit more leeway in semver terms.

Reply:
Maybe https://github.com/jaredpalmer/tsdx/issues/365 will lead to something soon, then we could put the 1.2.0 out as a "types & tree-shaking" update. 
Nice side-effect of this is that we aren't bound by tsdx 0.10 any more.